reporters_db/data/laws.json

```diff
@@ -5486,12 +5530,13 @@
                 "1 U.S.C. § 1",
                 "1 U.S.C.A. § 1",
                 "1 U.S.C.S. § 1",
-                "1 U.S.C.U. § 1"
+                "1 U.S.C.U. § 1",
+                "1 U.S.C. sec. 1"
             ],
             "jurisdiction": "United States",
             "name": "United States Code; United States Code Annotated; United States Code Service; Gould’s United States Code Unannotated",
             "regexes": [
-                "(?P<title>\\d+) $reporter § $law_section"
+                "(?P<title>\\d+) $reporter (?:§|Sec|sec|Section|section)[§|s]?\\.? $law_section"
```

Comment:
what do you think about changing?

`(?P<title>\\d+) $reporter (?:§|Sec|sec|Section|section)[§|s]?\\.? $law_section`
to
`(?P<title>\d+) $reporter (?:§|[Ss](ec)?(tion)?\.?) $law_section`

I'm certainly not a regex expert.  Either way I think § is also in there twice.

Reply:
Thanks @flooie. The reason I put § in twice (and added the extra groups you removed) was do that we will detect plurals. When citing to multiple sections, the correct form is §§ 2-4 or §§ 1, 2, and 3.  Your other changes, however, look good to me. The final code would be:
`(?P<title>\d+) $reporter (?:§|[Ss](ec)?(tion)?)[§|s]?\.? $law_section`

Reply:
Got it.  Also - I had never seen `(?:` before.  I went and read about it - but didnt quite understand what it does here for us.

Reply:
I suppose we do not need it if all of our capture group selections use group titles rather than iterators, but it defines a non-capture group, meaning that matches are not stored in the match groups. It is used where all matches are returned and used for future operations, or else where matches are seleced based on their iterator number rather than their title.

Reply:
Can I just take a second look at your regex pattern before we agree on it.

Reply:
Of course.

Reply:
I think this might be better

`(§§?)|[Ss](ec)?(tions?)?\.?`

the one you suggested . I think would capture weird anomalies like `S§.`

Reply:
Also I'm not sure what your preference is but I do not like putting spaces in regex. I prefer to use the whitespace command (\s) in case there is a nonbreaking space or something similar, and I usually put * or + so that multiple spaces match and, in the former case, a missing space will also match. For example:
```
(?P<title>\d+)\s+$reporter\s+(?:§|[Ss](ec)?(tion)?)[§|s]?\.?\s*$law_section
```
requires some whitespace between the title and the reporter name and between the reporter name and the section identifier, but it merely permits white space between the section identifier and the law_section. This means §1 will match.

Reply:
I'm pro whitespace.  I think it makes more sense to cleanup the text via preprocessing that can handle those situations ... and we have tools for that in eyecite to do some stuff like that. I dont know if @jcushman or @mlissner have thoughts on this though.

Reply:
Oh and one final note. The current regex treats §§ as two matches rather than one match of the plural option. If we'd rather it be one match, we can change the section_marker regex from 
`
(?:§|[Ss](ec)?(tion)?)[§|s]?\.?
`
to
`
(?:§§|§|[Ss](ec)?(tion)?s?\.?)
`

Reply:
I usually use `\s+` to match whitespace too, since it's more universal, but I'm not sure it matters in this application. We're not dealing with newlines or whatever that'd cause trouble, I don't think.

Reply:
> I think this might be better
> 
> `(§§?)|[Ss](ec)?(tions?)?\.?`
> 
> the one you suggested . I think would capture weird anomalies like `S§.`

How about this?

`(?:\s§§|§|[Ss](?:ec)(?:tion)?s?\.?\s)`

The one you suggested would match S or s whenever it appears alone or in a word (e.g. "case" matches). Making "ec" required solves this, but it will still match words that include sec or section (e.g. "intersect" or "Sectarian"), which is why I required the match to be preceded and succeeded by whitespace. We could use lookaheads and lookbehinds for this, but it would slow things down and, if the plan is to use the section_marker in many of the law citations, the speed may be significant. Yours also won't match the plural Secs. because you made the s option in "tions".

Finally, I made the internal groups non-capturing in case anybody has a use case in which the pull matches by iterators/position. If we are using the section_marker, people will not be able to look at the regex to see how many match groups it creates in laws.json, so it could really throw them off.

<img width="716" alt="Screen Shot 2022-09-02 at 4 13 46 PM" src="https://user-images.githubusercontent.com/25389306/188229303-af158a6e-2141-4d9f-927e-18a3867ad365.png">

Reply:
Are you saying that `S` alone without the period or even `s.` would be wrong?

Reply:
Also - this regex pattern isn't in a vacuum right.  The fact that it allows s. alone isn't an issue if thats correct because it will have to be s. alone surrounded by 1 U.S.C. ..... 203

Reply:
Hm. It is no more wrong than Sec., but I had been assuming it would create too many false positives if we matched s alone. Maybe if we require a space on either side it would be fine since it will only be used in regexes where it has to meet the other criteria (e.g. it needs to follow a reporter or statute or something). Come to think of it, reading § as S is probably a common OCR error too. Maybe we can try it with your new tool and see if it creates too many false positives.

Matching S. or s. makes sense to me as long as we require a space before.

Reply:
as soon as you update your code - it should automatically run the tests and we will know for sure. This all sounds good to me.

Reply:
thanks for going back and forth on this with me.

Reply:
> Also - this regex pattern isn't in a vacuum right. The fact that it allows s. alone isn't an issue if thats correct because it will have to be s. alone surrounded by 1 U.S.C. ..... 203

Yes exactly. I think we should try it with your new tool and see what happens.

In that case it would be `(?:§§|§|\s[Ss](?:ec)?(?:tion)?s?\.?)`. This requires a space before anything except the section symbol, but not after in case an OCR error renders §1 as S1 or something like that.

Reply:
> it should automatically run the tests and we will know for sure

I hesitate to ask, but I've been wondering: Does the Eyecite Report find and account for law citations or just case law citations?

Reply:
`(?:ec)?` why do we need this.

Reply:
and I stand by ... maybe wrongly that `(§§?)` works.

Reply:
im not sure the non capturing group matters, but now that I re-read it, its fine by me to throw in.

Reply:
> and I stand by ... maybe wrongly that `(§§?)` works.

You are right. I will use (?:(?:§§?)|\s[Ss](?:ec)?(?:tion)?s?\.?).

Reply:
All these non capturing groups feel redundant.  Maybe we can drop them all.

Reply:
> im not sure the non capturing group matters, but now that I re-read it, its fine by me to throw in.

Will do. I agree that it probably does not matter but, if that's right and it doesn't matter, then we should use non-capture groups because they are [faster and save memory](https://www.pythontutorial.net/python-regex/python-regex-non-capturing-group/). As I understand it, this is why non-capture groups are considered a best practice. Also, if anybody is using recorders-db and does use all matches or pull matches by iterator, using non-capture groups cannot mess them up but using capture groups could.

Reply:
Great.  That convinced me.

Reply:
Hmm I wonder if we can speed up eyecite with them.

Reply:
It'd be a decent test for your new benchmark tool, but I'll be the bad guy and place my bet that it'll be negligible. :)

That said, I'm with @bbernicker that using non-capturing groups is probably a best practice instead of capturing groups. They're just so ugly that I find I can't care very much. Readability vs Performance vs Exactness.

Reply:
Just out of curiosity, I just ran a quick test and tried matching 100,000,000 strings with and without non-capture groups using our section_marker regex. It turns out that the one which use capturing groups was faster. I was always taught that not saving the group contents freed up memory and made the code faster, but for whatever reason that does not seem to be true in this circumstance. 

So in the interest of readability, I will scrap them and, if anyone is using iterators to pull match groups, this will be a wake-up call to them to stop doing something so senseless.

Reply:
@flooie, sorry if I missed it, but did you check on this or have thoughts about it:

> I hesitate to ask, but I've been wondering: Does the Eyecite Report find and account for law citations or just case law citations?